Providing API to set HTTP2 max concurrent streams

kunalekawde opened the pull request:
Hello,
Added an API to set the max concurrent streams for HTTP2 connections. Request to please review.
Thanks

bagder commented:
Thanks a lot for your contribution! Some initial feedback and questions:

1. Try `make checksrc` locally first to find the nits the code style checker finds. Saves you time.
2. How about coming up with a name for the option that would also work for HTTP/3?
3. What happens if this is set when there are existing connections already setup? (goes to the documentation).
4. Any min/max values ? Users will need some guidance here. What happens if you set max to 20K or 128G?

kunalekawde commented:
@bagder , thanks for the feedback and some important points you shared, I shall check these. 
PS: This is my first git open source add process, please pardon for any miss and open for learning :)

bagder commented:
@kunalekawde, my comments are in no way a critique of your approach or the process you're following. They're only meant to help making this an even better PR and get it into a shape that we can end up merging. I really appreciate you helping us out. There's also no hurry, since this is a feature/change that we, according to our merge rules, can't merge into master until after the pending release anyway so we have a few weeks to polish it to perfection!

I'm honored you're dipping your toes into open source for real for the first time with us, and I hope we won't discourage you from doing more of it going forward! :smile:

bagder commented on lib/multi.c:
```diff
@@ -3043,3 +3046,8 @@ void Curl_multi_dump(struct Curl_multi *multi)
 }
 #endif
+
+uint32_t Curl_multi_max_http2_concurrent_streams(struct Curl_multi *multi)
```
inconveniently enough, we don't use `uint32_t` in our code (and this causes compiler error on some places). Stick to good old `unsigned int` and you'll get an unsigned 32 bit variable.

kunalekawde replied:
Sure thanks, I had moved it to long, following other examples. Shall check on this.

kunalekawde commented:
@bagder 
>>How about coming up with a name for the option that would also work for HTTP/3?
Yeah, I quick read through QUIC / HTTP3 and moved the name to max_concurrent_streams instead of max_http2_concurrent_streams. Basically removed http2 from it.

>>What happens if this is set when there are existing connections already setup? (goes to the >>documentation).
Yes, I've updated the documentation for default, min and max. But documentation needs to be updated with target libcurl version.

>>Any min/max values ? Users will need some guidance here. What happens if you set max to 20K or 128G?
I've put an upper limit on 2^32 -1 (referred other implementation limits)

kunalekawde commented:
@bagder , from the build logs, I don't see any issue with code, looks like timeout and FTP case failed. Any pointers to look at ?

jay commented:
CIs have arbitrary slowdowns which cause some of our tests to fail sometimes. I've restarted Travis and Cirrus CIs but not appveyor since I can't figure out how. (My appveyor for some reason shows jay/curl and not curl/curl even though when I go into appveyor settings I see it set to curl/curl so not sure what that's about I'll have to ask them.)

IlyaFinkelshteyn commented:
@jay answered [here](https://help.appveyor.com/discussions/questions/39344-why-cant-i-restart-a-pr#comment_47427836). After you added as a [Collaborator](https://www.appveyor.com/docs/team-setup/#collaborator), I would recommend `RE-RUN INCOMPLETE` button to save some time.

kunalekawde commented:
Thanks @jay and @IlyaFinkelshteyn for the support , so as I understand, I need to wait for appveyor to be restarted, post which I need to push some dummy change from my repo or the builds would run automatically ?

IlyaFinkelshteyn commented:
@kunalekawde pushing some dummy change will work, but it will restart all other CIs and all AppVeyor jobs. Instead someone who has permissions to AppVeyor project can press `RE-RUN INCOMPLETE` to restart just [single failed job](https://ci.appveyor.com/project/curlorg/curl/builds/25443474/job/3oc9j3f5pqvlrphy). CC @bagder

kunalekawde commented:
Thanks @IlyaFinkelshteyn , Looks like it was run as only coverage is failing, shall check this.

jay commented on include/curl/multi.h:
```diff
@@ -382,6 +382,9 @@ typedef enum {
   CINIT(PUSHDATA, OBJECTPOINT, 15),
 
+  /* maximum number of (pipelining) connections to one host */
+  CINIT(MAX_CONCURRENT_STREAMS, LONG, 16),
```
comment is wrong

kunalekawde replied:
Thanks, updated

jay commented on include/curl/multi.h:
```diff
@@ -434,6 +437,9 @@ typedef int (*curl_push_callback)(CURL *parent,
                                   void *userp);
 
+/* value for MAXIMUM CONCURRENT STREAMS upper limit */
+#define INITIAL_MAX_CONCURRENT_STREAMS ((1U << 31) - 1)
```
suggest DEFAULT instead of INITIAL like DEFAULT_MAX_CONCURRENT_STREAMS

jay replied:
nevermind I see why you called it INITIAL_MAX_CONCURRENT_STREAMS.

jay commented on lib/multi.c:
```diff
@@ -3043,3 +3051,11 @@ void Curl_multi_dump(struct Curl_multi *multi)
+{
+  return multi ?
+    (multi->max_concurrent_streams?
+    multi->max_concurrent_streams:100) : 0;
```
these lines are too broken up they can be longer up to 79

kunalekawde replied:
Thanks, updated.

jay commented on packages/OS400/curl.inc.in:
```diff
@@ -1821,6 +1821,8 @@
      d  CURLMOPT_MAX_PIPELINE_LENGTH...
      d                 c                   00008
+     d  CURLMOPT_MAX_CONCURRENT_STREAMS...
+     d                 c                   00009
```
this is wrong it has to be appended 
     d  CURLMOPT_MAX_CONCURRENT_STREAMS...
     d                 c                   00016

kunalekawde replied:
```
     d  CURLMOPT_PUSHDATA...
     d                 c                   10015
     d  CURLMOPT_MAX_CONCURRENT_STREAMS...
     d                 c                   10016
      *
      * Bitmask bits for CURLMOPT_PIPELING.
      *

```
Is this correct ?

jay commented on docs/examples/http2-download.c:
```diff
@@ -203,6 +203,8 @@ int main(int argc, char **argv)
     curl_multi_add_handle(multi_handle, trans[i].easy);
   }
 
+  curl_multi_setopt(multi_handle, CURLMOPT_MAX_CONCURRENT_STREAMS, 120L);
```
this is not needed in the boilerplate example. also I had intended my comments to be submitted all at once but I never clicked start review.. sorry about that!

kunalekawde replied:
ok sure, shall remove this

bagder commented on docs/libcurl/opts/CURLMOPT_MAX_CONCURRENT_STREAMS.3:
```diff
@@ -20,18 +20,18 @@
 support.
-
+Valid values range from 1 to 2147483647 (2^31 - 1) and defaults to 100.
 This option is for the multi handle's use only.
```
I don't think this needs to say it is for multi handles only. The option is only for a single function and that function is for a multi handle.

I think the description needs to say something about this value being a request and that there's no guarantee that the set value will actually be used.

kunalekawde replied:
> 
> 
> I don't think this needs to say it is for multi handles only. The option is only for a single function and that function is for a multi handle.
> 
Agree
> I think the description needs to say something about this value being a request and that there's no guarantee that the set value will actually be used.
Actually upstream nghttp2 doesn't mention something like this. May be I can only update that its the max value and would be dependent on transfer traffic rate.

bagder commented on docs/libcurl/symbols-in-versions:
```diff
@@ -318,7 +318,7 @@ CURLMOPT_MAXCONNECTS            7.16.3
 CURLMOPT_MAX_PIPELINE_LENGTH    7.30.0
 CURLMOPT_MAX_TOTAL_CONNECTIONS  7.30.0
-CURLMOPT_MAX_HTTP2_CONCURRENT_STREAMS  7.64.3
+CURLMOPT_MAX_CONCURRENT_STREAMS  7.64.3
```
`7.66.0` as things look now!

bagder commented on lib/multi.c:
```diff
@@ -3047,10 +3051,10 @@ void Curl_multi_dump(struct Curl_multi *multi)
-    (multi->max_http2_concurrent_streams?
-    multi->max_http2_concurrent_streams:100) : 0;
+    (multi->max_concurrent_streams?
+    multi->max_concurrent_streams:100) : 0;
```
`multi->max_concurrent_streams` is a `long` and this returns a `size_t` ? On some platforms like Windows, they are not of equal size.

kunalekawde replied:
Its the same for max_host_connections, max_total_connections. May be I can type cast it to size_t, which would adjust according to architecture used.

bagder commented on lib/multi.c:
```diff
@@ -2625,8 +2625,12 @@ CURLMcode curl_multi_setopt(struct Curl_multi *multi,
+    {
+      long streams = va_arg(param, long);
+      multi->max_concurrent_streams = (streams>INITIAL_MAX_CONCURRENT_STREAMS)?
+          INITIAL_MAX_CONCURRENT_STREAMS : streams;
```
This will allow negative values to be set. I presume that isn't intended? And zero.

kunalekawde replied:
Agree, updated

kunalekawde commented:
@bagder , @jay , any help /pointers on this ?

for travis build:configure: error: --with-quiche was specified but could not find quiche pkg-config file.
for appveyor: not sure if its testcase test1501

bagder commented:
If you rebase your branch the quiche problem will go away as it is already fixed.

Note that we strongly recommend rebase, **not** merging master into your branch, as it makes your work much easier to review.

kunalekawde commented:
Thanks @bagder 
I tried following in my local clone:

```
$ git rebase origin/master
Current branch master is up to date.

$ git status
On branch master
Your branch is up-to-date with 'origin/master'.

nothing to commit, working tree clean

[kunal_ekawde@cni-builder-15-15-34 curl]$ git pull
Already up-to-date.
```

There are no changes pulled in.

bagder commented:
1. make sure your origin/master is in sync with the upstream's (curl/curl) origin/master
2. checkout your local feature branch and rebase that to your origin/master 
3. force push to your local feature branch

kunalekawde commented:
Thanks @bagder , apologies for delay as I was on vacation.
I had to play around git & not sure if its as expected. Somehow I ended up creating a branch.

```
$ git push --set-upstream origin max-concurrent-streams -f
Total 0 (delta 0), reused 0 (delta 0)
remote:
remote: Create a pull request for 'max-concurrent-streams' on GitHub by visiting:
remote:      https://github.com/kunalekawde/curl/pull/new/max-concurrent-streams
remote:
To https://github.com/kunalekawde/curl.git
 * [new branch]          max-concurrent-streams -> max-concurrent-streams
Branch max-concurrent-streams set up to track remote branch max-concurrent-streams from origin.

$ git branch -vv
  master                 046e1fbd6 [origin/master] Merge branch 'master' of https://github.com/kunalekawde/curl
* max-concurrent-streams 046e1fbd6 [origin/max-concurrent-streams] Merge branch 'master' of https://github.com/kunalekawde/curl
```
 
Do you think this is correct and any minimum step required to finish  this ? Or if there is some alternate way if you could suggest.

kunalekawde commented:
Sorry as I'm not able to get correct git commands may be, only way I see here is to create a draft pull as suggested so that changes / diff can be seen and absorbed. Please let me know, so I can atleast close this pull request.

jay commented:
It's because the branch of the PR is set to master. It's not good to work in master, typically you make a branch for the PR which it looks like you did later, max-concurrent-streams. There might be a way to change the branch associated with this PR but I think it would be easier to just open a new one and reference this one.

kunalekawde commented:
Thanks @jay , let me do that.
